diff --git a/gzip.c b/gzip.c
index 913fafe..95250f4 100644
--- a/gzip.c
+++ b/gzip.c
@@ -97,6 +97,11 @@ static char const *const license_msg[] = {
 #  include <utimens.h>
 #endif
 
+#ifdef __MVS__
+# include "zos_extra.h"
+# include <zos.h>
+#endif
+
 #ifndef MAX_PATH_LEN
 #  define MAX_PATH_LEN   1024 /* max pathname length */
 #endif
@@ -185,7 +190,9 @@ static int part_nb;          /* number of parts in .gz file */
 static char *env;            /* contents of GZIP env variable */
 static char const *z_suffix; /* default suffix (can be set with --suffix) */
 static size_t z_len;         /* strlen(z_suffix) */
-
+#ifdef __MVS__
+       int zedc = 1;         /* use IBM Z hardware compression (zEDC) by default*/
+#endif
 /* The original timestamp (modification time).  If the original is
    unknown, TIME_STAMP.tv_nsec is negative.  If the original is
    greater than struct timespec range, TIME_STAMP is the maximal
@@ -220,6 +227,53 @@ unsigned inptr;            /* index of next byte to be processed in inbuf */
 unsigned outcnt;           /* bytes in output buffer */
 int rsync = 0;             /* make rsyncable chunks */
 
+#ifdef __MVS__
+// Credit to Rocket Software, this is adapted from their source code distribution
+
+zos_extra_field_t zos_extra_field;
+
+int get_file_attr(int fd) {
+  if( verbose  ) fprintf( stderr, "%s( fd:%d ): ", __FUNCTION__, fd );
+  memset( &zos_extra_field, 0, sizeof(zos_extra_field_t) );
+  memcpy( &zos_extra_field.sign, ZOS_SIGN, 3 );
+
+  struct stat st;
+  if( fstat( fd, &st ) != -1 )
+  {
+      zos_extra_field.ft_ccsid = st.st_tag.ft_ccsid;
+      if( verbose )
+          fprintf( stderr, "ccsid: %d (0x%04X)\n",
+                    zos_extra_field.ft_ccsid, zos_extra_field.ft_ccsid );
+      return 0;
+  }
+
+  int e = errno;
+  memset( &zos_extra_field, 0, sizeof(zos_extra_field_t) );
+  fprintf( stderr, "\n%s: %s(): can't get file attributes for '%s'\n%d - %s\n",
+           program_name, __FUNCTION__, ofname, e, strerror(e) );
+  gzip_error(NULL);
+  return 1;
+}
+
+int restore_file_attr(int fd) {
+  if( verbose ) fprintf( stderr, "%s( fd:%d ): ", __FUNCTION__, fd );
+  if( memcmp( zos_extra_field.sign, ZOS_SIGN, 3 ) != 0 ) {
+      if (verbose) {
+          fprintf( stderr, "\n%s: %s(): extra field block has no signature 'zos'. "
+                            "skip...\n", program_name, __FUNCTION__ );
+          return 0;
+      }
+  }
+
+  if (verbose)
+    fprintf( stderr, "\n%s: %s(): changing ccsid: %d\n",  program_name, __FUNCTION__, zos_extra_field.ft_ccsid  );
+  if (zos_extra_field.ft_ccsid == 0) 
+    __setfdccsid( fd, zos_extra_field.ft_ccsid );
+  else
+    __chgfdccsid( fd, zos_extra_field.ft_ccsid );
+}
+#endif
+
 static int handled_sig[] =
   {
     /* SIGINT must be first, as 'foreground' depends on it.  */
@@ -249,6 +303,9 @@ enum
   PRESUME_INPUT_TTY_OPTION = CHAR_MAX + 1,
   RSYNCABLE_OPTION,
   SYNCHRONOUS_OPTION,
+#ifdef  __MVS__
+  NO_ZEDC_OPTION,
+#endif
 };
 
 static char const shortopts[] = "ab:cdfhH?klLmMnNqrS:tvVZ123456789";
@@ -284,6 +341,9 @@ static const struct option longopts[] =
     {"lzw",        0, 0, 'Z'}, /* make output compatible with old compress */
     {"bits",       1, 0, 'b'}, /* max number of bits per code (implies -Z) */
     {"rsyncable",  0, 0, RSYNCABLE_OPTION}, /* make rsync-friendly archive */
+#ifdef	__MVS__
+    {"nz",         0, 0, NO_ZEDC_OPTION}, /* Dont use IBM Z hardware compression (zEDC) */
+#endif
     { 0, 0, 0, 0 }
 };
 
@@ -365,6 +425,9 @@ help ()
  "  -t, --test        test compressed file integrity",
  "  -v, --verbose     verbose mode",
  "  -V, --version     display version number",
+#ifdef __MVS__
+ "      --nz          Dont use IBM Z hardware compression/decompression (zEDC) by default",
+#endif
  "  -1, --fast        compress faster",
  "  -9, --best        compress better",
  "",
@@ -383,7 +446,6 @@ static void
 license ()
 {
     char const *const *p = license_msg;
-
     printf ("%s %s\n", program_name, Version);
     while (*p) printf ("%s\n", *p++);
 }
@@ -567,7 +629,11 @@ int main (int argc, char **argv)
         case '5':  case '6':  case '7':  case '8':  case '9':
             level = optc - '0';
             break;
-
+#ifdef  __MVS__
+	case NO_ZEDC_OPTION:
+	    zedc = 0;
+	    break;
+#endif
         default:
             try_help ();
         }
@@ -733,6 +799,13 @@ treat_stdin ()
     ifd = STDIN_FILENO;
     stdin_was_read = true;
 
+#ifdef __MVS__
+    __disableautocvt(ifd);
+    if (!decompress)
+      get_file_attr(ifd);
+    __disableautocvt(STDOUT_FILENO);
+#endif
+
     if (decompress) {
         method = get_method(ifd);
         if (method < 0) {
@@ -914,6 +987,12 @@ treat_file (char *iname)
     clear_bufs(); /* clear input and output buffers */
     part_nb = 0;
 
+#ifdef __MVS__
+    if (!decompress)
+      get_file_attr(ifd);
+    __disableautocvt(ifd);
+#endif
+
     if (decompress) {
         method = get_method(ifd); /* updates ofname if original given */
         if (method < 0) {
@@ -937,6 +1016,17 @@ treat_file (char *iname)
                     program_name, ifname, ofname);
         }
     }
+
+#ifdef __MVS__
+     __disableautocvt(ofd);
+    if (!decompress) {
+      if(!to_stdout) 
+        __setfdbinary(ofd);
+    } else if (!to_stdout) {
+      restore_file_attr(ofd);  
+    }
+#endif
+
     /* Keep the name even if not truncated except with --no-name: */
     if (!save_orig_name) save_orig_name = !no_name;
 
@@ -1547,6 +1637,17 @@ get_method (int in)
             uch lenbuf[2];
             unsigned int len = lenbuf[0] = get_byte ();
             len |= (lenbuf[1] = get_byte ()) << 8;
+#ifdef __MVS__
+            uch *p = (uch*)&zos_extra_field;
+            if( verbose  ) fprintf( stderr, "In zos_get_extra_field()");
+            for( int i = 0; i < len && i < sizeof(zos_extra_field_t); i++ )
+            {
+                *p = get_byte();
+                p++;
+            }
+            if (flags & HEADER_CRC)
+              updcrc ((uch*)&zos_extra_field, sizeof(zos_extra_field_t));
+#else
             if (verbose) {
                 fprintf(stderr,"%s: %s: extra field of %u bytes ignored\n",
                         program_name, ifname, len);
@@ -1554,6 +1655,7 @@ get_method (int in)
             if (flags & HEADER_CRC)
               updcrc (lenbuf, 2);
             discard_input_bytes (len, flags);
+#endif
         }
 
         /* Get original file name if it was truncated */
