diff --git a/gzip.h b/gzip.h
index 855a88b..2c08afc 100644
--- a/gzip.h
+++ b/gzip.h
@@ -192,6 +192,9 @@ extern int level;          /* compression level */
 extern int test;           /* check .z file integrity */
 extern int to_stdout;      /* output to stdout (-c) */
 extern int save_orig_name; /* set if original name must be saved */
+#ifdef __MVS__
+extern int zedc;           /* use IBM Z hardware compression (zEDC) */
+#endif
 
 #define get_byte()  (inptr < insize ? inbuf[inptr++] : fill_inbuf(0))
 #define try_byte()  (inptr < insize ? inbuf[inptr++] : fill_inbuf(1))
@@ -330,3 +333,9 @@ extern int gzip_inflate (void);
 extern int dfltcc_deflate (int pack_level);
 extern int dfltcc_inflate (void);
 #endif
+
+#ifdef __MVS__
+extern int is_zedc_enabled (int decompress);
+extern int zedc_deflate (int pack_level);
+extern int zedc_inflate (void);
+#endif
