diff --git a/zip.c b/zip.c
index eb42982..4487a81 100644
--- a/zip.c
+++ b/zip.c
@@ -21,6 +21,10 @@
 
 #include "tailor.h"
 #include "gzip.h"
+#ifdef __MVS__
+#include "zos_extra.h"
+#include <zos.h>
+#endif
 
 off_t header_bytes;   /* number of bytes in gzip header */
 
@@ -55,6 +59,13 @@ zip (int in, int out)
     if (save_orig_name) {
         flags |= ORIG_NAME;
     }
+
+#ifdef __MVS__
+    zos_extra_field_t *zz = zos_get_extra_field();
+    if( memcmp( zz->sign, ZOS_SIGN, 3 ) == 0 )
+        flags |= EXTRA_FIELD; // data for restore a file tag in zos
+#endif
+
     put_byte(flags);         /* general flags */
     if (time_stamp.tv_nsec < 0)
       stamp = 0;
@@ -71,17 +82,34 @@ zip (int in, int out)
 
     /* Write deflated file to zip file */
     updcrc (NULL, 0);
-
+#ifdef __MVS__
+    if(!zedc)
+    {
+#endif
     bi_init(out);
     ct_init(&attr, &method);
     if (level == 1)
       deflate_flags |= FAST;
     else if (level == 9)
       deflate_flags |= SLOW;
+#ifdef __MVS__
+    }
+#endif
 
     put_byte((uch)deflate_flags); /* extra flags */
     put_byte(OS_CODE);            /* OS identifier */
 
+#ifdef __MVS__
+    if( (flags & EXTRA_FIELD) != 0 )
+    {
+        ush len = sizeof( zos_extra_field_t );
+        put_short( len );
+        uch *p = (uch*)zz;
+        for( int i = 0; i < len; i++ )
+            put_byte( p[i] );
+    }
+#endif
+
     if (save_orig_name) {
         char *p = gzip_base_name (ifname); /* Don't save the directory part. */
         do {
@@ -93,6 +121,14 @@ zip (int in, int out)
 #ifdef IBM_Z_DFLTCC
     dfltcc_deflate (level);
 #else
+#ifdef __MVS__
+    if (zedc)
+    {
+      flush_outbuf();
+      zedc_deflate (level);
+    }
+    else
+#endif
     gzip_deflate (level);
 #endif
 
